package/ota_daemon/src/ota_daemon.c

```diff
@@ -0,0 +1,529 @@
+      case OPT_ID_URL: {
+        opt_url = optarg;
+      }
+      break;
+
+      default: {
+        puts("Invalid option");
+        return -1;
+      }
+      break;
+    }
+  }
+
+  if (op_mode == OP_MODE_OTA_CLIENT && opt_url == NULL) {
+    opt_url = OTA_DEFAULT_URL;
+    printf("Using default URL: %s\n", opt_url);
+  }
+
+  return 0;
+}
+
+static void ota_sig_handler(int signum, siginfo_t *info, void *ucontext)
+{
+  (void)ucontext;
+  piksi_log(LOG_DEBUG,
+            "%s: received signal: %d, sender: %d",
+            __FUNCTION__,
+            signum,
+            info->si_pid);
+  libnetwork_shutdown(NETWORK_TYPE_OTA);
+}
+
+static int configure_libnetwork(network_context_t *ctx, int fd, const char *url)
+{
+  network_status_t status = NETWORK_STATUS_SUCCESS;
+
+  if ((status = libnetwork_set_url(ctx, url)) != NETWORK_STATUS_SUCCESS) {
+    piksi_log(LOG_ERR|LOG_SBP, "Failed configuring libnetwork url: %d", status);
+    return 1;
+  }
+
+  if ((status = libnetwork_set_fd(ctx, fd)) != NETWORK_STATUS_SUCCESS) {
+    piksi_log(LOG_ERR|LOG_SBP, "Failed configuring libnetwork fd: %d", status);
+    return 1;
+  }
+
+  if ((status = libnetwork_set_debug(ctx, false)) != NETWORK_STATUS_SUCCESS) {
+    piksi_log(LOG_ERR|LOG_SBP, "Failed configuring libnetwork debug: %d", status);
+    return 1;
+  }
+
+  status = libnetwork_set_continuous(ctx, false);
+  if (status != NETWORK_STATUS_SUCCESS) {
+    piksi_log(LOG_ERR|LOG_SBP,
+              "Failed configuring libnetwork continuous fd: %d",
+              status);
+    return 1;
+  }
+
+  return 0;
+}
+
+static int ota_upgrade() {
+  int ret = 0;
+
+  const char *upgrade_cmd =
+    "sh -c 'set -o pipefail; "
+    "sudo upgrade_tool --debug "OTA_IMAGE_FILE_PATH" && "
+    "sudo /etc/init.d/do_sbp_msg_reset'";
+
+  runit_config_t cfg = (runit_config_t) {
+    .service_dir    = RUNIT_SERVICE_DIR,
+    .service_name   = "upgrade_tool",
+    .command_line   = upgrade_cmd,
+    .finish_command = NULL,
+    .restart        = false,
+  };
+
+  ret = start_runit_service(&cfg);
+
+  if (ret) {
+    piksi_log(LOG_ERR|LOG_SBP, "Failed to start upgrade_tool: %d", ret);
+  }
+
+  return ret;
+}
+
+static int ota_sha256sum(const char *expected) {
+  const char *cmd = "sha256sum";
+  char *const argv[3] = { "sha256sum", OTA_IMAGE_FILE_PATH, NULL };
+  char sha[65] = {0};
+  int ret = run_with_stdin_file(NULL, cmd, argv, sha, sizeof(sha));
+
+  if (ret) {
+    piksi_log(LOG_ERR|LOG_SBP, "sha256sum error (%d), errno: %d", ret, errno);
+    return ret;
+  }
+
+  ret = strcmp(sha, expected);
+
+  if (ret) {
+    piksi_log(LOG_ERR, "SHA256 mismatch");
+    piksi_log(LOG_ERR, "%s", sha);
+    piksi_log(LOG_ERR, "%s", expected);
+  }
+
+  return ret;
+}
+
+/* TODO: check that token includes only digits */
+#define CHECK_TOKEN(version) do {                                   \
+  if (token) {                                                      \
+    version = atoi(token);                                          \
```

Comment:
`strtoul` doc says that it returns zero for non-valid conversion and since `"0"` is a valid version string I think it won't work for us. I added helper `str_digits_only`.